src/server/modules/bulk/services/BulkService.ts

```diff
@@ -0,0 +1,86 @@
+import { injectable } from 'inversify'
+import paparse from 'papaparse'
+import * as interfaces from '../interfaces/BulkService'
+
+import {
+  bulkUploadMaxNum,
+  bulkUploadRandomStrLength,
+  ogHostname,
+} from '../../../config'
+import { BULK_UPLOAD_HEADER } from '../../../../shared/constants'
+import { BulkUrlMapping } from '../../../repositories/types'
+import * as validators from '../../../../shared/util/validation'
+import generateShortUrl from '../../../util/url'
+
+const BULK_UPLOAD_RANDOM_STRING_LENGTH = bulkUploadRandomStrLength
+const BULK_UPLOAD_MAX_NUM = bulkUploadMaxNum
+
+@injectable()
+export class BulkService implements interfaces.BulkService {
+  parseCsv: (dataString: string) => interfaces.CSVSchema = (dataString) => {
+    const schema = {
+      rows: 0,
+      isValid: true,
+      longUrls: [],
+    } as interfaces.CSVSchema
+
+    paparse.parse(dataString, {
+      skipEmptyLines: false,
+      delimiter: ',',
+      step(step, parser) {
+        const rowData = step.data as string[]
+        let validRow = true
+        if (schema.rows === 0) {
+          // if header is invalid
+          if (rowData[0] !== BULK_UPLOAD_HEADER) {
+            schema.isValid = false
+            parser.abort()
+          }
+        } else {
+          const acceptableLinkCount = schema.rows <= BULK_UPLOAD_MAX_NUM + 1 // rows include header
+          const onlyOneColumn = rowData.length === 1
+          const isNotBlacklisted = !validators.isBlacklisted(rowData[0])
```

Comment:
nit: I think we can extract this variable rowData[0] out

Reply:
[6c1df00](https://github.com/opengovsg/GoGovSG/pull/1993/commits/6c1df00e95d231082c7fe109f6f12f749a8f82a4)